database/tests/authentication_tests/test_tableau_api_auth.py

```diff
@@ -0,0 +1,127 @@
+import requests
+from unittest.mock import MagicMock
+
+from api.tableau_api.base import TableauApiView, AuthenticationFailed, PermissionDenied
+from app import app
+from database.study_models import Study, DeviceSettings
+from database.tests.authentication_tests.django_flask_hybrid_test_framework import HybridTest
+from database.tests.authentication_tests.testing_constants import TEST_USERNAME, TEST_PASSWORD, BASE_URL, \
+    TEST_STUDY_NAME, TEST_STUDY_ENCRYPTION_KEY
+from database.security_models import ApiKey
+from database.user_models import Researcher, StudyRelation
+from time import sleep
+
+
+class TableauApiAuthTests(HybridTest):
+    """
+    Test methods of the api authentication system
+    """
+
+    def setup(self, researcher=True, apikey=True, study=True):
+        if apikey and not researcher:
+            raise Exception("invalid setup criteria")
+        if researcher:
+            self.researcher = Researcher.create_with_password(username=TEST_USERNAME, password=TEST_PASSWORD)
+        if apikey:
+            api_key = ApiKey.generate(self.researcher, has_tableau_api_permissions=True)
+            self.api_key_public = api_key.access_key_id
+            self.api_key_private = api_key.access_key_secret_plaintext
+        if study:
+            self.study = Study.create_with_object_id(device_settings=DeviceSettings(), encryption_key=TEST_STUDY_ENCRYPTION_KEY, name=TEST_STUDY_NAME)
+            if researcher:
+                StudyRelation(study=self.study, researcher=self.researcher, relationship="researcher").save()
+
+
+    @staticmethod
+    def login(session=None):
+        if session is None:
+            session = requests.Session()
+        session.post(BASE_URL + "/validate_login", data={'username': TEST_USERNAME, 'password': TEST_PASSWORD})
```

Comment:
url_for requires the application context, which I did not find any sensible way to mock as in the request context. I am trying a different approach by storing that data in the constants file which should at least make it easy to change to and find mistakes in.

Reply:
You may be able to do something like:
```
class HybridTestCase():
    app = flask_app

    def __init__(self, *args, *kwargs):
        super().__init__(*args, **kwargs)
        self.app.config["SERVER_NAME"] = f"{HOST}:{PORT}"

    def url_for(self, endpoint, **values):
        with self.app.app_context():
            return url_for(endpoint, **values)
```
![image](https://user-images.githubusercontent.com/1454606/89925336-42961000-dbd1-11ea-98e9-96b3e8b2e54b.png)

There is an argument for hardcoding urls, but I think that only really matters for externally-facing APIs (so for internal urls like login, `url_for` is great; for the Tableau API endpoint, it's good to hard code since we want to ensure we don't accidentally change the endpoint since external users depend on that).

Reply:
Thanks for the help with that. I had tried something similar to your first block of code but ran into errors I couldn't figure out. The screenshot code worked great however, and this definitely seems like the better solution.